### PR TITLE
Reconnection and Ping/Pong

### PR DESCRIPTION
**Pull request by RaduDumitriu-Dev**
- separated reconnection behaviour from other pull request as a separate PR
- adds ping/ponging to current Colyseus room. pings occur every 2 seconds, and after 3 failed consecutive pings the session disconnects
- on disconnect or ping failure, session will attempt to reconnect to the session over several attempts every 5 seconds. after 60 attempts, attempts will cease and playthrough will be fully disconnected.

### CODE REVIEW

File: Scripts/Playthrough.cs [hunk 1]
```diff
@@ -374,6 +341,208 @@ private bool IsConnected()
...
+
+            while (room == null)
+            {
+                // Try to reconnect
+                _reconnectionTryCount++;
+
+                if (_reconnectionTryCount <= MAXIMUM_RECONNECTION_ATTEMPTS)
+                {
+                    _client.Settings.useSecureProtocol = true;
+                    var t = _client.Reconnect(_room.Id, _room.SessionId);
```
Comment (bensalilijames) [inline Scripts/Playthrough.cs:483]: usually when reconnecting, we can get away with just using `Reconnect()` as it is here, which works as long as the playthrough 'room' still exists on the server

however sometimes, especially when a server updates, the playthrough room is dropped entirely and the client needs to request to re-create the room

we can test if that's the case because `_client.Reconnect()` will throw an error which looks something like `room "ROOM_ID" not found`

in this scenario we need to call `client.JoinOrCreate()` in the same way we do initially, it may be easier to just call `CreateClientAndJoinRoom` again

so in a nutshell:

- keep trying to `_client.Reconnect()`
- if `_client.Reconnect()` fails with an error matching the string `room "ROOM_ID" not found` then instead...
- keep trying to `_client.JoinOrCreate()`

(also rename `t` to something more descriptive like `reconnectionTask`!)

reference in JS SDK: https://github.com/charisma-ai/charisma-sdk-js/blob/ac3c243d5e79be654d91aa7d03363e3d19721e5a/src/Playthrough.ts#L270

File: Scripts/Playthrough.cs [hunk 2]
```diff
@@ -374,6 +341,208 @@ private bool IsConnected()
...
+
+                        _room = room;
+                        AssignRoomCallbacks();
+                        SetConnectionState(ConnectionState.Connected);
+                        yield break;
+                    }
+                    catch (Exception e)
+                    {
+                        Logger.LogWarning($"Failed to reconnect - exception: {e.Message}");
+                    }
```
Comment (bensalilijames) [inline Scripts/Playthrough.cs:501]: curious why there's a `try/catch` block here, what bit of code could throw an error in the `try` block?

File: Scripts/Playthrough.cs [hunk 3]
```diff
@@ -374,6 +341,208 @@ private bool IsConnected()
...
+        private bool IsReconnecting()
+        {
+            return _connectionState == ConnectionState.Reconnecting;
+        }
+
+        private IEnumerator Reconnect()
+        {
+            ColyseusRoom<object> room = null;
+
+            while (room == null)
```
Comment (bensalilijames) [inline Scripts/Playthrough.cs:475]: maybe `while(_room == null)` instead rather than creating a new variable `room`?

**Review by bensalilijames — CHANGES_REQUESTED** [review]

File: Scripts/Playthrough.cs (hunk 2 above)
Comment (RaduDumitriu-Dev) [reply Scripts/Playthrough.cs:501]: The reconnecting Task `t` can fail here if no internet is available. I added that try and catch since that seemed like the best way to account for a Task failing to complete in C# and ensuring we don't run any "connection succesful" functions after a failed reconnect

File: Scripts/Playthrough.cs (hunk 3 above)
Comment (RaduDumitriu-Dev) [reply Scripts/Playthrough.cs:475]: I added this because we're technically still referencing the original `_room` when issuing the Reconnect task via the ID and Session. I can store those IDs locally in separate variables ofc, but seemed more convenient to have this

File: Scripts/Playthrough.cs [hunk 4]
```diff
@@ -374,6 +341,235 @@ private bool IsConnected()
             return _connectionState == ConnectionState.Connected;
         }
 
+        private bool HasBeenDisconnected()
+        {
+            return _connectionState == ConnectionState.Disconnected;
+        }
+
+        private async Task CreateClientAndJoinRoom()
```
Comment (bensalilijames) [inline Scripts/Playthrough.cs:349]: since this method doesn't return anything, should this be `void`?
```suggestion
        private async void CreateClientAndJoinRoom()
```

File: Scripts/Playthrough.cs [hunk 5]
```diff
@@ -374,6 +341,235 @@ private bool IsConnected()
...
+                    }
+                    else
+                    {
+                        // Try just reconnecting otherwise.
+                        reconnectionTask = _client.Reconnect(_room.Id, _room.SessionId);
+                    }
+
+                    yield return new WaitUntil(() => reconnectionTask.IsCompleted);
+
+                    try
```
Comment (bensalilijames) [inline Scripts/Playthrough.cs:500]: I'm not sure if this `try/catch` block does anything as there's no code within the `try` block that could throw an error

I'm not so good at coroutines/async but instead I think we want something like:

```c#
if (reconnectionTask.Status != TaskStatus.Faulted) {
  // happy case
} else {
  // sad case
  // we can get the exception from the task (note it seems to be wrapped in `AggregateException`)
  reconnectionTask.Exception;
}
```

ref: https://learn.microsoft.com/en-us/dotnet/standard/parallel-programming/exception-handling-task-parallel-library

File: Scripts/Playthrough.cs [hunk 6]
```diff
@@ -374,6 +341,235 @@ private bool IsConnected()
...
+                    {
+                        room = reconnectionTask.Result;
+
+                        // If the room is succesfully generated and no error is catched
+                        // code will progress here and re-assign the _room variable.
+                        _room = room;
+                        AssignRoomCallbacks();
+                        SetConnectionState(ConnectionState.Connected);
+                        Logger.Log($"Succesfully reconnected.");
```
Comment (bensalilijames) [inline Scripts/Playthrough.cs:509]:
```suggestion
                        Logger.Log($"Successfully reconnected.");
```

**Review by bensalilijames — CHANGES_REQUESTED** [review]

File: Scripts/Playthrough.cs (hunk 5 above)
Comment (RaduDumitriu-Dev) [reply Scripts/Playthrough.cs]: Thanks, that definitely works better! I can confirm the try/catch was working - accessing the task.Result while in a try still counts as an exception and it would trigger catch in my breakpoint tests. But this is a lot more readable and sensible

**Review by bensalilijames — APPROVED** [review]
looks good! have tested locally and works both when connection is dropped, and also when the server initiates the disconnect itself 💯

**Comment by RaduDumitriu-Dev** [comment]
Thanks Ben! Merging this now
